### DISCUSSION TITLE
Large number of FS mounts left behind (approx 49k)

### DISCUSSION THREAD

**Discussion by HemangLavana** [concourse/concourse#6851] [Help & Support]
I am using concourse 7.1.0 on RHEL8. It seems that there are large number (~49k) of overlay FS mounts that are created on the worker node and NOT cleaned up. Apparently, this causes the linux box to hang and a reboot is required to make it usable. I have enabled use of "containerd". 

I am sure I must have mis-configured something to cause this. How can I go about debugging this?

Example:
 ```
[root@sjc-***-u*** ~]# mount | grep concourse | wc -l
49936

[root@sjc-***-u*** ~]# cat /etc/redhat-release
 Red Hat Enterprise Linux release 8.1 (Ootpa)

[root@sjc-***-u*** ~]# mount | grep concourse | more
/dev/mapper/mpatha1 on /nobackup/ngdevx/concourse/worker/volumes/live/00034918-3d5b-48e1-7ddd-82d60500e7f3/volume type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
/dev/mapper/mpatha1 on /nobackup/ngdevx/concourse/worker/volumes/live/0012724e-738e-40b5-4565-298121f81b10/volume type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
/dev/mapper/mpatha1 on /nobackup/ngdevx/concourse/worker/volumes/live/00139c11-510a-4722-58ef-6a56d3be7d08/volume type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
...
```

**Comment by taylorsilva** [comment]
Overlay mounts are managed by the baggageclaim component on the worker: https://github.com/concourse/baggageclaim/
49k is a crazy amount of volumes to have mounted! I just checked one of our larger deployments and they all have <1k overlay mounts per worker.

Everything under `worker/volumes/live/` means it is currently mounted in a container (at least that's the intention of `live`). Did you find anything interesting in the logs? grepping for `baggageclaim` is a good idea. Hopefully you can find a good error message to point you in the right direction.

**Reply by HemangLavana** [reply]
I checked the number of live mounts and they do seem to be 49k (very strange):

$ ls /nobackup/.../concourse/worker/volumes/live/  | wc -l
49935
$

I grepped for the baggaeclaim errors in the stdout/stderr of worker node startup script. Here are some sample type of messages that I found:

{"timestamp":"2021-03-31T19:27:48.263017857Z","level":"error","source":"baggageclaim","message":"baggageclaim.api.volume-server.list-volumes.failed-to-encode","data":{"error":"write tcp 127.0.0.1:7788-\u003e127.0.0.1:36150: write: broken pipe","session":"3.1.2"}}
...
{"timestamp":"2021-03-31T19:30:50.025149083Z","level":"info","source":"baggageclaim","message":"baggageclaim.api.volume-server.get-volume.volume-not-found","data":{"session":"3.1.76","volume":"cb8b61da-ce99-4a42-5a7f-7594d48638df"}}
...
and then eventually, I start getting these messages even though none of the local filesystems are full (which I check with df -h):

{"timestamp":"2021-04-06T18:34:50.201075573Z","level":"error","source":"baggageclaim","message":"baggageclaim.api.volume-serve
r.create-volume-async.create-volume.failed-to-materialize-strategy","data":{"error":"no space left on device","handle":"92ab24

https://discuss.concourse-ci.org/t/failed-to-create-volume-error-when-running-pipelines/1471 seems to suggest that I should set CONCOURSE_BAGGAGECLAIM_DRIVER=naive.  I will be trying this  option out.

Any hints on what else can I do?
5b-eb3e-4b8e-4aa4-71886ce89c9e","session":"3.1.527163.1"}}

**Reply by taylorsilva** [reply]
The naive strategy will definitely work but it will use up more disk space and increase task initialization. naive just does `cp` on volumes ([literally](https://github.com/concourse/baggageclaim/blob/cc2c3140310612ea2c5d3503a3a781d9b0be9228/volume/driver/naive.go#L24)) whenever it needs a copy for a new container. The default strategy, overlay, is the recommended one to go with. It saves disk space and is fast.

Interesting that you're getting `no space left on device`; how close to full are your disks usually? I'd guess that the disk filled up for a moment before Concourse GC'd some volumes, thus freeing up disk space again.

Is your worker crashing/restarting/recreated regularly? The `volume-not-found` error could be normal if the worker restarted with an empty disk but with the same name. In that situation, when a worker disappears and another returns with the same name, the web will expect it to have the same volumes as the original worker. This is fine though, as the web will remove them from memory when it's given the `volume-not-found` error.

Lastly, when a worker restarts, if there are any volumes on disk baggagclaim will try to recover them. If the workers configuration hasn't changed this works pretty well and all volumes should be recovered properly. Baggageclaim does everything in-memory, so the filesystem structure is the source of truth for baggageclaim; it doesn't rely on a db.

Another thing you could try, if you do `ls -l /nobackup/ngdevx/concourse/worker/volumes/live/`, what's the oldest volume? That would help tell you if volumes aren't being GC'd.

**Reply by HemangLavana** [reply]
The naive strategy is not helping either, so I have switched back to overlay.

I looked at the worker with 49k live mounts which is no longer attached to concourse web node. The ls -lt /nobackup/ngdevx/concourse/worker/volumes/live/ returns the timestamps from Mar 13th to Apr 6th. 

Also,  it seems unlikely that the diskspace is getting filled up as I  have 5TB diskspace. On the above node with 49k live mounts, it currently only shows 13% full (even inodes are plenty):

$ df -h /nobackup/
Filesystem           Size  Used Avail Use% Mounted on
/dev/mapper/mpatha1  5.0T  613G  4.4T  13% /nobackup
$ df -ih /nobackup/
Filesystem          Inodes IUsed IFree IUse% Mounted on
/dev/mapper/mpatha1   503M   27M  476M    6% /nobackup
$

Moreover, the worker node is no longer connected to the web node and doesn't show up in the fly -t xx workers list. However, there are still 49k mounts on this node:

$ mount | fgrep concourse |wc -l
49935
$

Also, I am wondering if there is some concourse pipleine or job that is badly configured and creating these checks all the time. Is there some metric available to check what could be causing this?

Here are some more errors that I came across in the logs:

# this only occurs initially near worker startup
{"timestamp":"2021-04-15T02:47:06.250464795Z","level":"error","source":"worker","message":"worker.beacon-runner.beacon.forward-conn.failed-to-dial","data":{"addr":"127.0.0.1:7777","error":"dial tcp 127.0.0.1:7777: connect: connection refused","network":"tcp","session":"4.1.4"}}

# this one occurs several times
{"timestamp":"2021-04-15T02:47:27.779785102Z","level":"error","source":"worker","message":"worker.garden.garden-server.run.failed","data":{"error":"get container: container \"a770067c-934c-41b3-5c9c-892de3522fe5\" in namespace \"concourse\": not found","handle":"a770067c-934c-41b3-5c9c-892de3522fe5","session":"1.2.2"}}

# this one is occurring the most
{"timestamp":"2021-04-15T02:54:37.660618031Z","level":"error","source":"worker","message":"worker.garden.garden-server.attach.failed","data":{"error":"load proc: no running process found: process does not exist resource: not found","handle":"c01769ec-a11c-480c-6b0c-e168593f6213","session":"1.2.210"}}

PS: I also upgraded to 7.2.0 today, so will know in a few days if it has fixed any of the above issues.

**Reply by taylorsilva** [reply]
> Also, I am wondering if there is some concourse pipleine or job that is badly configured and creating these checks all the time.

This is my current thought as well. It is possible to tie the volumes on a worker back to a pipeline by looking at the output from `fly volumes` and `fly containers`. I forget the exact steps how but it is possible!

The first and third error you posted are noisy errors, they aren't indicative of any real problems on you worker. The second one could be a symptom of the real problem, but hard to know based on that one error.

Having volumes that are a month old "feels" odd to me. Most volumes eventually stopped being used, usually due to new versions being found or containers being deleted. 

Pick one of the volumes from the end of March, so middle of the group. The handle for the volume is the folder name after `/nobackup/ngdevx/concourse/worker/volumes/live/<volume-handle>`. Connect to the Concourse db and search for the handle in the volumes table: `select * from volumes where handle='<volume-handle>'`. I want to know if Concourse is still tracking this volume or not. This should also help you figure out what team and pipeline it's from. A rogue pipeline still seems like the most likely cause. I wonder if there's some long-running task running somewhere and multiple builds of this task are building up?

**Reply by HemangLavana** [reply]
I checked one of the live volume that was created about 13hrs back and is on a connected worker node:

```
postgres-ci=# select * from volumes where handle='7b8cc186-6d28-48d3-74c5-f8ba0c48a599' ;
   id    |                handle                |  worker_name   | path | container_id | team_id |  state  | parent_id | paren
t_state | worker_base_resource_type_id | worker_resource_cache_id | worker_task_cache_id | worker_resource_certs_id | missing_
since | worker_artifact_id 
---------+--------------------------------------+----------------+------+--------------+---------+---------+-----------+------
--------+------------------------------+--------------------------+----------------------+--------------------------+---------
------+--------------------
 1276112 | 7b8cc186-6d28-48d3-74c5-f8ba0c48a599 | sjc-devx-u3028 |      |              |         | created |           |      
        |                          959 |                          |                      |                          |         
      |                   
(1 row)
```

Most of the fields are empty. When I look for this handle in the "fly -t x volumes --detail --json" output, it contains many references for it and here's one same json data for it:
```json
  {
    "id": "ae38a973-7b20-4289-52ec-548151993794",
    "worker_name": "sjc-devx-u3028",
    "type": "container",
    "container_handle": "",
    "path": "/",
    "parent_handle": "7b8cc186-6d28-48d3-74c5-f8ba0c48a599",
    "resource_type": null,
    "base_resource_type": null,
    "pipeline_id": 0,
    "pipeline_name": "",
    "pipeline_instance_vars": null,
    "job_name": "",
    "step_name": ""
  }
```

Next, I looked at another random live volume from about 14hrs ago.  It seems to have no data in db:

```
postgres-ci=# select * from volumes where handle='8ba2b663-59ef-44e1-60f1-a0bd7f71b9d' ;
 id | handle | worker_name | path | container_id | team_id | state | parent_id | parent_state | worker_base_resource_type_id |
 worker_resource_cache_id | worker_task_cache_id | worker_resource_certs_id | missing_since | worker_artifact_id 
----+--------+-------------+------+--------------+---------+-------+-----------+--------------+------------------------------+
--------------------------+----------------------+--------------------------+---------------+--------------------
(0 rows)
```

and here's the json output for it:
```json
{
    "id": "8ba2b663-59ef-44e1-60f1-a0bd7f71b9d9",
    "worker_name": "sjc-devx-u3028",
    "type": "container",
    "container_handle": "",
    "path": "/",
    "parent_handle": "7b8cc186-6d28-48d3-74c5-f8ba0c48a599",
    "resource_type": null,
    "base_resource_type": null,
    "pipeline_id": 0,
    "pipeline_name": "",
    "pipeline_instance_vars": null,
    "job_name": "",
    "step_name": ""
  }
```

I looked up the workers logs to see what was it doing:

```
...
concourse-worker-sjc-devx-u3028.log:{"timestamp":"2021-04-15T02:50:34.289283735Z","level":"debug","source":"baggageclaim","message":"baggageclaim.api.volume-server.create-volume-async-check.done","data":{"handle":"8ba2b663-59ef-44e1-60f1-a0bd7f71b9d9","session":"3.1.105"}}
logs/run-concourse-worker-sjc-devx-u3028.log:{"timestamp":"2021-04-15T02:50:34.290539124Z","level":"debug","source":"baggageclaim","message":"baggageclaim.api.volume-server.create-volume-async-cancel.start","data":{"handle":"8ba2b663-59ef-44e1-60f1-a0bd7f71b9d9","session":"3.1.106"}}
logs/run-concourse-worker-sjc-devx-u3028.log:{"timestamp":"2021-04-15T02:50:34.290573919Z","level":"debug","source":"baggageclaim","message":"baggageclaim.api.volume-server.create-volume-async-cancel.done","data":{"handle":"8ba2b663-59ef-44e1-60f1-a0bd7f71b9d9","session":"3.1.106"}}
logs/run-concourse-worker-sjc-devx-u3028.log:{"timestamp":"2021-04-15T02:51:46.604913034Z","level":"info","source":"worker","message":"worker.garden.garden-server.create.created","data":{"request":{"Handle":"2843011e-53b8-4390-61dd-653fa72838cd","GraceTime":0,"RootFSPath":"raw:///nobackup/ngdevx/concourse/worker/volumes/live/8ba2b663-59ef-44e1-60f1-a0bd7f71b9d9/volume","BindMounts":[{"src_path":"/nobackup/ngdevx/concourse/worker/volumes/live/87e12e90-0d8d-4dc2-6a7e-52ef195b3954/volume","dst_path":"/scratch","mode":1},{"src_path":"/nobackup/ngdevx/concourse/worker/volumes/live/e8217a90-5f4a-4257-4b26-b92d7eda8299/volume","dst_path":"/tmp/build/get","mode":1}],"Network":"","Privileged":true,"Limits":{"bandwidth_limits":{},"cpu_limits":{},"disk_limits":{},"memory_limits":{},"pid_limits":{}}},"session":"1.2.79"}}
logs/run-concourse-worker-sjc-devx-u3028.log:{"timestamp":"2021-04-15T02:51:46.624607858Z","level":"debug","source":"baggageclaim","message":"baggageclaim.api.volume-server.get-volume.start","data":{"session":"3.1.201","volume":"8ba2b663-59ef-44e1-60f1-a0bd7f71b9d9"}}
logs/run-concourse-worker-sjc-devx-u3028.log:{"timestamp":"2021-04-15T02:51:46.624695575Z","level":"debug","source":"baggageclaim","message":"baggageclaim.api.volume-server.get-volume.done","data":{"session":"3.1.201","volume":"8ba2b663-59ef-44e1-60f1-a0bd7f71b9d9"}}
```

Does it provide any clues on what could be wrong.

**Reply by taylorsilva** [reply]
okay that's weird, they're all type `container` which I think means they're the base image for a container. But the two volumes you shared don't have a `container_handle`? If you take the `id` for one of those volumes, does it appear as the `parent_handle` for another volume?
Also, go the other way, take the `parent_handle` for one of those volumes and look up it's parent volume (`parent_handle`).

Are most of these volumes of type `container`?

**Reply by taylorsilva** [reply]
Just looked through the output of `fly volumes` on our ci and all the volumes with no `container_handle` were for resource, resource-type, or task-cache which is expected. I couldn't find any volumes with type container and an empty container_handle.

**Reply by HemangLavana** [reply]
In my case, most of volumes have empty container_handle (3119) and only  about 81 have a value for container_handle
 for it. See below (volumes.json file contains the output of fly -t x volumes --details --json):

$ fgrep container_handle volumes.json |fgrep  \"\" |wc -l
3119
$ fgrep container_handle volumes.json |fgrep  -v \"\" |wc -l
81
$ fgrep \"type\" volumes.json |fgrep  container |wc -l
3160
$

The 3rd query shows that most of the volumes are of type: container and majority of those have empty container_handles.

I was looking at fly volumes output of the "main" team where I don't have any pipelines configured. But it still returned large number of volumes. So I suspect it could be different based on whether the pipelines are public or not and whether a given team has access to all the volumes or not.

**Comment by HemangLavana** [comment]
I checked multiple volumes from the disconnected node: (Mar 18, Mar 27th and Apr 6th) and none of them exist in concourse db. But the 49k+ mounts are still existing on that node.

I will check  the volumes for other live worker nodes once it increases and see what happen to them.

Also, do "fly -t t1 volumes" and "fly -t t2 volumes" return the list of volumes associated with the specified target only or all the volumes? I am getting different results with different targets: for one I am getting about 2.5k volumes and for another about 2.9k volumes. And when I diff the two outputs, it has about 1.8k diff lines. So it seems some are common volumes reported and some are different.

Note that I have about 47 pipelines configured on this instance by various team members, and when I look at the main concourse dashboard, none of them seem to be running (any boxes blinking). Perhaps some are doing lot of periodic checks. I will focus on debugging ones created by me  to see if I can find the root  cause.

**Reply by taylorsilva** [reply]
Looking at the code, it appears it will only return volumes for the specific team you're targeting. That explains the diff.

**Comment by HemangLavana** [comment]
I was looking at one of the worker nodes which is not connected to concourse web node. It has about 49k live mounts and if I try to re-connect as worker node, it fails with "out of disk-space" error message. But when I check the various partitions, they all seem to have plenty of free space. See below:

[ scripts]$ sudo concourse worker 1> logs/run-concourse-worker-sjc-devx-u3021.log 2> logs/run-concourse-worker-sjc-devx-u3021.err

[ scripts]$ cat logs/run-concourse-worker-sjc-devx-u3021.log 
{"timestamp":"2021-04-20T17:55:09.714396003Z","level":"info","source":"baggageclaim","message":"baggageclaim.using-driver","data":{"driver":"overlay"}}
{"timestamp":"2021-04-20T17:55:09.962287052Z","level":"error","source":"baggageclaim","message":"baggageclaim.failed-to-recover-volume-driver","data":{"error":"recover bind mount: no space left on device"}}
[ scripts]$ df -h /nobackup/ngdevx/concourse/worker/volumes
Filesystem           Size  Used Avail Use% Mounted on
/dev/mapper/mpatha1  5.0T  600G  4.4T  12% /nobackup
[ scripts]$ df -ih /nobackup/ngdevx/concourse/worker/volumes
Filesystem          Inodes IUsed IFree IUse% Mounted on
/dev/mapper/mpatha1   503M   26M  477M    6% /nobackup
[ scripts]$ df -h /
Filesystem               Size  Used Avail Use% Mounted on
/dev/mapper/redhat-root   20G  4.7G   16G  24% /
[ scripts]$ df -h /tmp
Filesystem      Size  Used Avail Use% Mounted on
tmpfs           126G  107M  126G   1% /tmp
[ scripts]$ df -h /var/log
Filesystem              Size  Used Avail Use% Mounted on
/dev/mapper/redhat-var   20G  1.4G   19G   7% /var
[ scripts]$ mount |fgrep concourse |wc -l
49940
[ scripts]$ ls /nobackup/ngdevx/concourse/worker/volumes/live |wc -l
49941
[ scripts]$

Any hints on how to find where it is having diskspace issues.

**Comment by link2xt** [comment]
I have a similar problem:
```
# mount | grep concourse | wc -l
47027
```

Worker does not start anymore:
```
May 06 20:51:09 b1 systemd[1]: Stopped Concourse CI worker process.
May 06 20:51:09 b1 systemd[1]: Started Concourse CI worker process.
May 06 20:51:09 b1 concourse[2812176]: {"timestamp":"2024-05-06T18:51:09.370574468Z","level":"info","source":"baggageclaim","message":"baggageclaim.using-driver","data":{"driver":"overlay"}}
May 06 20:51:09 b1 concourse[2812176]: {"timestamp":"2024-05-06T18:51:09.496745991Z","level":"error","source":"baggageclaim","message":"baggageclaim.failed-to-recover-volume-driver","data":{"error":"recover bind mount: no space left on device"}}
May 06 20:51:09 b1 concourse[2812176]: error: recover bind mount: no space left on device
May 06 20:51:09 b1 systemd[1]: concourse-worker.service: Main process exited, code=exited, status=1/FAILURE
May 06 20:51:09 b1 systemd[1]: concourse-worker.service: Failed with result 'exit-code'.
May 06 20:51:09 b1 systemd[1]: concourse-worker.service: Scheduled restart job, restart counter is at 5.
May 06 20:51:09 b1 systemd[1]: Stopped Concourse CI worker process.
May 06 20:51:09 b1 systemd[1]: concourse-worker.service: Start request repeated too quickly.
May 06 20:51:09 b1 systemd[1]: concourse-worker.service: Failed with result 'exit-code'.
May 06 20:51:09 b1 systemd[1]: Failed to start Concourse CI worker process.
```

Running Concourse CI v7.11.0 with `CONCOURSE_RUNTIME=containerd`.

This seems to make Docker commands very slow as they have to go through all these overlays. Also `prometheus-node-exporter` is very busy dealing with all these mounts.
